File: app/src/main/java/com/billbao/sample/teams/sort/SortType.java
package com.billbao.sample.teams.sort;

import com.billbao.data.model.Team;
import java.util.Comparator;

/**
 * Use enum type to hold the different comparators.
 */
public enum SortType {
    FULLNAME(new TeamFullNameComparator()),
    WINS(new TeamWinsComparator()),
    LOSSES(new TeamLossesComparator());

    private Comparator<Team> mComparator;

    SortType(Comparator<Team> comparator) {
        mComparator = comparator;
    }

    public Comparator<Team> comparator() {
        return mComparator;
    }

    private static class TeamFullNameComparator implements Comparator<Team> {
        @Override public int compare(Team first, Team second) {
            return first.getFullName().compareToIgnoreCase(second.getFullName());
        }
    }

    private static class TeamWinsComparator implements Comparator<Team> {
        @Override public int compare(Team first, Team second) {
            return first.getWins().compareTo(second.getWins());
        }
    }

    private static class TeamLossesComparator implements Comparator<Team> {
        @Override public int compare(Team first, Team second) {
            return first.getLosses().compareTo(second.getLosses());
        }
    }
}

File: app/src/main/java/com/billbao/sample/teams/sort/SortType.kt
package com.billbao.sample.teams.sort

import com.billbao.data.model.Team
import java.util.Comparator

/**
 * Use enum type to hold the different comparators.
 */
enum class SortType private constructor(private val mComparator: Comparator<Team>) {
    FULLNAME(TeamFullNameComparator()),
    WINS(TeamWinsComparator()),
    LOSSES(TeamLossesComparator());

    fun comparator(): Comparator<Team> {
        return mComparator
    }

    private class TeamFullNameComparator : Comparator<Team> {
        override fun compare(first: Team, second: Team): Int {
            return first.fullName.compareTo(second.fullName, ignoreCase = true)
        }
    }

    private class TeamWinsComparator : Comparator<Team> {
        override fun compare(first: Team, second: Team): Int {
            return first.wins!!.compareTo(second.wins)
        }
    }

    private class TeamLossesComparator : Comparator<Team> {
        override fun compare(first: Team, second: Team): Int {
            return first.losses!!.compareTo(second.losses)
        }
    }
}
